python/src/scipp/compat/mantid.py

```diff
@@ -193,17 +193,119 @@ def validate_and_get_unit(unit):
+    c = np.dot(a, b)
+    s = np.linalg.norm(v)
+    if s == 0:
+        return np.eye(3)
+    kmat = np.array([[0, -v[2], v[1]], [v[2], 0, -v[0]], [-v[1], v[0], 0]])
+    rotation_matrix = np.eye(3) + kmat + kmat.dot(kmat) * ((1 - c) / (s**2))
+    return rotation_matrix
+
+
+def init_pos(ws, source_pos, sample_pos):
+    from scipp import Dim
+    import numpy as np
     spec_info = ws.spectrumInfo()
-    for i in range(nHist):
-        if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+    det_info = ws.detectorInfo()
+    total_detectors = spec_info.detectorCount()
+    non_detectors = sum(not spec.hasDetectors for spec in spec_info)
+    det_pos = np.zeros([total_detectors, 3 + 1])
+    no_detector_spectrum = np.full((non_detectors, 3 + 1), np.nan)
+
+    if sample_pos and source_pos:
+        act_beam = (sample_pos - source_pos).values
+        rot = rotation_matrix_from_vectors(act_beam, [0, 0, 1])
+        inv_rot = rot.transpose()
+    else:
+        rot = np.eye(3)
+        inv_rot = rot
+
+    idx, idx_orphaned = 0, 0
+    for i, spec in enumerate(spec_info):
+        if spec.hasDetectors:
+            definition = spec_info.getSpectrumDefinition(i)
+            n_dets = len(definition)
+            for j in range(n_dets):
+                det_idx = definition[j][0]
+                p = det_info.position(det_idx)
+                _to_array(det_pos, idx, p, i)
+                idx += 1
         else:
-            pos[i, :] = [np.nan, np.nan, np.nan]
+            no_detector_spectrum[idx_orphaned, 0] = i
+            idx_orphaned += 1
+    det_pos[:, 1:] = det_pos[:, 1:].dot(rot)
+    spherical_ = _to_spherical(det_pos)
+
+    pos_d = sc.Dataset()
+    pos_d.coords["spectrum_idx"] = sc.Variable([Dim.X],
+                                               values=spherical_[:, 0])
+    pos_d["r"] = sc.Variable([Dim.X], values=spherical_[:, 1], unit=sc.units.m)
+    pos_d["t"] = sc.Variable([Dim.X],
+                             values=spherical_[:, 2],
+                             unit=sc.units.rad)
+    pos_d["p"] = sc.Variable([Dim.X],
+                             values=spherical_[:, 3],
+                             unit=sc.units.rad)
+
+    averaged = sc.groupby(pos_d, "spectrum_idx").mean(Dim.X)
+
+    no_detector_ds = sc.Dataset()
+    no_detector_ds.coords["spectrum_idx"] = sc.Variable(
+        ["spectrum_idx"],
+        values=np.arange(len(no_detector_spectrum)),
+        dtype=np.float)
+    no_detector_ds["r"] = sc.Variable(
+        ["spectrum_idx"],
+        values=np.array([np.nan] * len(no_detector_spectrum)),
+        unit=sc.units.m)
+    no_detector_ds["t"] = sc.Variable(
+        ["spectrum_idx"],
+        values=np.array([np.nan] * len(no_detector_spectrum)),
+        unit=sc.units.rad)
+    no_detector_ds["p"] = sc.Variable(
+        ["spectrum_idx"],
+        values=np.array([np.nan] * len(no_detector_spectrum)),
+        unit=sc.units.rad)
```

Comment:
All the missing det handling seems overly complicated. Can't you just use a fake Id such as `-1` for missing dets, then `groupby`, and finally ignore the "-1" group?

Reply:
Yes, that would slightly simplify the grouping. We still have to account for spectra with no detectors. We need to know what the indexes of those spectra are so that we can return a properly sorted result.

Reply:
Why not something like:

1. Create table (dataset) with all detector information
2. loop over all spectra, write spectrum index for all detectors
3. `groupby(..., `spectrum-index`, bins=np.arange(-0.5, nspec + 0.5, 1.0))` (using `groupby` with `bins` to get NAN for spectra, instead of not having output.
4. Convert back to position and copy output arrays.